File: Agora-Online-Chorus-iOS/AgoraChorusDemo/AgoraChorusDemo/MainViewController.m
//
//  MainViewController.m
//  AgoraChorusDemo
//
//  Created by GongYuhua on 2016/9/12.
//  Copyright © 2016年 Agora. All rights reserved.
//

#import "MainViewController.h"
#import "SettingsViewController.h"
#import "LiveRoomViewController.h"
#import <AVFoundation/AVFoundation.h>
@interface MainViewController () <SettingsVCDelegate, LiveRoomVCDelegate, UITextFieldDelegate>
@property (weak, nonatomic) IBOutlet UITextField *roomNameTextField;
@property (weak, nonatomic) IBOutlet UIView *popoverSourceView;
@property (strong, nonatomic) AVAudioPlayer *player;
@property (assign, nonatomic) AgoraVideoProfile videoProfile;
@end

@implementation MainViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.videoProfile = AgoraVideoProfilePortrait480P;//AgoraRtc_VideoProfile_480P;
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    NSString *segueId = segue.identifier;
    
    if ([segueId isEqualToString:@"mainToSettings"]) {
        SettingsViewController *settingsVC = segue.destinationViewController;
        settingsVC.videoProfile = self.videoProfile;
        settingsVC.delegate = self;
    } else if ([segueId isEqualToString:@"mainToLive"]) {
        LiveRoomViewController *liveVC = segue.destinationViewController;
        liveVC.roomName = self.roomNameTextField.text;
        liveVC.videoProfile = self.videoProfile;
        liveVC.params = [sender integerValue];
        liveVC.delegate = self;
    }
}

- (void)showRoleSelection {
    UIAlertController *sheet = [UIAlertController alertControllerWithTitle:nil message:nil preferredStyle:UIAlertControllerStyleActionSheet];
    UIAlertAction *broadcaster = [UIAlertAction actionWithTitle:@"加入房间" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [self joinWithRole:1];
    }];
    [sheet addAction:broadcaster];
    UIAlertAction *cancel = [UIAlertAction actionWithTitle:@"cancel" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
    }];
    [sheet addAction:cancel];
    [sheet popoverPresentationController].sourceView = self.popoverSourceView;
    [sheet popoverPresentationController].permittedArrowDirections = UIPopoverArrowDirectionUp;
    [self presentViewController:sheet animated:YES completion:nil];
}

- (void)joinWithRole:(NSInteger )parms {
    
    [self performSegueWithIdentifier:@"mainToLive" sender:@(parms)];
}

- (void)settingsVC:(SettingsViewController *)settingsVC didSelectProfile:(AgoraVideoProfile)profile {
    self.videoProfile = profile;
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (void)liveVCNeedClose:(LiveRoomViewController *)liveVC {
    [self.navigationController popViewControllerAnimated:YES];
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    if (textField.text.length) {
        [self showRoleSelection];
    }
    
    return YES;
}
@end

File: Agora-Online-Chorus-iOS/AgoraOnlineChorus/Controllers/MainViewController.swift
//
//  MainViewController.swift
//  AgoraOnlineChorus
//
//  Created by ZhangJi on 2019/3/21.
//  Copyright © 2019 ZhangJi. All rights reserved.
//

import UIKit

enum EnterType {
    case singer
    case playback
}

class MainViewController: UIViewController {
    
    @IBOutlet weak var roomNameTextField: UITextField!
    @IBOutlet weak var popoverSourceView: UIButton!
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        guard let segueId = segue.identifier, let roomName = sender as? String else {
                return
        }
        
        switch segueId {
        case "mainToRoom":
            let roomVC = segue.destination as! RoomViewController
            roomVC.roomName = roomName
            roomVC.delegate = self
        case "mainToPlayback":
            let playbackVC = segue.destination as! PlaybackViewController
            playbackVC.roomName = roomName
            playbackVC.delegate = self
        default:
            break
        }
        
    }
    
    @IBAction func doRoomNameTextFieldEditing(_ sender: UITextField) {
        if let text = sender.text , !text.isEmpty {
            let legalString = MediaCharacter.updateToLegalMediaString(from: text)
            sender.text = legalString
        }
    }
    
    @IBAction func doJoinPressed(_ sender: UIButton) {
        showRoleSelection()
    }
}

private extension MainViewController {
    func showRoleSelection() {
        let sheet = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
        let singer = UIAlertAction(title: "As Singer", style: .default) { [weak self] _ in
            guard let strongSelf = self else {return}
            strongSelf.enter(roomName: strongSelf.roomNameTextField.text, type: .singer)
        }
        let playback = UIAlertAction(title: "Playback", style: .default) { [weak self] _ in
            guard let strongSelf = self else {return}
            strongSelf.enter(roomName: strongSelf.roomNameTextField.text, type: .playback)
        }
        let cancel = UIAlertAction(title: "Cancel", style: .cancel, handler: nil)
        sheet.addAction(singer)
        sheet.addAction(playback)
        sheet.addAction(cancel)
        sheet.popoverPresentationController?.sourceView = popoverSourceView
        sheet.popoverPresentationController?.permittedArrowDirections = .up
        present(sheet, animated: true, completion: nil)
    }
    
    func enter(roomName: String?, type: EnterType) {
        guard let roomName = roomName , !roomName.isEmpty else {
            return
        }
        switch type {
        case .singer:
            performSegue(withIdentifier: "mainToRoom", sender: roomName)
        case .playback:
            performSegue(withIdentifier: "mainToPlayback", sender: roomName)
        }
    }
}

extension MainViewController: RoomVCDelegate, PlaybackVCDelegate {
    func playbackVCNeedClose(_ roomVC: PlaybackViewController) {
        dismiss(animated: true, completion: nil)
    }
    
    func roomVCNeedClose(_ roomVC: RoomViewController) {
        dismiss(animated: true, completion: nil)
    }
}

extension MainViewController: UITextFieldDelegate {
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        showRoleSelection()
        return true
    }
}
